### ISSUE TITLE
Vectorized utility functions: abs2, unit phasor, others?

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#745]
#### Description
It would be useful to have an efficient function for getting magnitude-squared (power) from real or complex variables.  This came up in offline discussion of #742 , where `abs` is doing unnecessary work if we actually want power anyway.

This could be efficiently accomplished by using `numba.vectorize` to make a ufunc out of `x * x.conj()`.

**Comment by bmcfee** [comment]
While we're at it, it would also be nice to add an optimized helper function `util.phase` get the unit phasor representation of phase as currently returned by `magphase`.  This is a one-liner `np.exp(1.j * np.angle(x))`, but numba can (should be able to, at least) optimize this to eliminate intermediate storage.

**Comment by lostanlen** [comment]
This discussion is relevant: https://github.com/numpy/numpy/issues/3994

**Comment by bmcfee** [comment]
It'd be great if numpy solves this for us upstream. :grin:

**Comment by lostanlen** [comment]
Absolutely. I would recommend waiting at least a few months and see if a plan for `np.abs2` emerges in the numpy community. `librosa.abs2` would be a nice-to-have but definitely not a must have. It might save a few percent points at runtime, but I doubt it will make a big difference across all platforms.

**Comment by lostanlen** [comment]
the new reference point on the numpy repository is: https://github.com/numpy/numpy/issues/13179

**Comment by bmcfee** [comment]
My preference is still to leave this ball in numpy's court.  When they do implement the ufuncs mentioned in the issue you linked, we can do a pass through the codebase and see if things can be accelerated.

**Comment by lostanlen** [comment]
update: we have a PR! 🎉 
https://github.com/scipy/scipy/pull/15390

**Comment by bmcfee** [comment]
Scipy PR appears to be languishing.  However, the following bit of numba code seems to get me a nice boost:

```python
@numba.vectorize(['float32(complex64)', 'float64(complex128)'], identity=0)
def abs2(x):
    return x.real**2 + x.imag**2
```

Silly little benchmark (precision, array shape, running time), averaged over 100 reps each:

```python
sizes = [(1000,), (100, 100), (100, 1000), (2, 100, 10000)]
times_nb = []
times_np = []

for precision in [np.complex64, np.complex128]:
    for size in sizes:
        x = (np.random.randn(*size) + 1j * np.random.randn(*size)).astype(precision)
    
        t = timeit.timeit('abs2(x)', number=100, globals={'abs2': abs2, 'x': x})
        times_nb.append([precision, size, t])
        print('Numba: ', times_nb[-1])
        
        t = timeit.timeit('x.real**2 + x.imag**2', number=100, globals={'x': x})
        times_np.append([precision, size, t])
        print('Numpy: ', times_np[-1])
```

Results:
```
Numba:  [<class 'numpy.complex64'>, (1000,), 0.0009355060028610751]
Numpy:  [<class 'numpy.complex64'>, (1000,), 0.0005685449868906289]
Numba:  [<class 'numpy.complex64'>, (100, 100), 0.0008048400050029159]
Numpy:  [<class 'numpy.complex64'>, (100, 100), 0.001272291992790997]
Numba:  [<class 'numpy.complex64'>, (100, 1000), 0.008314680992043577]
Numpy:  [<class 'numpy.complex64'>, (100, 1000), 0.009603373997379094]
Numba:  [<class 'numpy.complex64'>, (2, 100, 10000), 0.19375343499996234]
Numpy:  [<class 'numpy.complex64'>, (2, 100, 10000), 0.41134571099246386]
Numba:  [<class 'numpy.complex128'>, (1000,), 8.470799366477877e-05]
Numpy:  [<class 'numpy.complex128'>, (1000,), 0.0002940880076494068]
Numba:  [<class 'numpy.complex128'>, (100, 100), 0.00047369500680360943]
Numpy:  [<class 'numpy.complex128'>, (100, 100), 0.0017795669991755858]
Numba:  [<class 'numpy.complex128'>, (100, 1000), 0.005280600002151914]
Numpy:  [<class 'numpy.complex128'>, (100, 1000), 0.015400140997371636]
Numba:  [<class 'numpy.complex128'>, (2, 100, 10000), 0.29070613300427794]
Numpy:  [<class 'numpy.complex128'>, (2, 100, 10000), 0.8062816750025377]
```

Seems like a pretty good boost to me.  Memory consumption should also be reduced for large arrays, since temporary storage can be optimized out by the compiler.

At risk of reversing my prior opinion, I think this is worth including and not waiting for numpy/scipy to implement for us.

It might also be worth looking into other simple/common operations that could be similarly accerlated.  One that comes to mind is mapping angles to complex points (this came up during griffin-lim updates in #1514 ) where we can either go in one shot via exp, or in two shots by cos/sin.  Seems like a good candidate for vectorization.

**Comment by bmcfee** [comment]
And just because I can't help myself, here's the phase mapping function:

```python
@numba.vectorize(['complex64(float32)', 'complex128(float64)'], identity=1+0j)
def phasor(x):
    return np.cos(x) + 1j * np.sin(x)
```
Benchmark code (vs numpy using trig functions and numpy using exp):
```python
sizes = [(1000,), (100, 100), (100, 1000), (1000, 1000), (2, 100, 10000)]
times_nb = []
times_nptrig = []
times_npexp = []

for precision in [np.float32, np.float64]:
    for size in sizes:
        x = np.random.randn(*size).astype(precision)
    
        t = timeit.timeit('phasor(x)', number=100, globals={'phasor': abs2, 'x': x})
        times_nb.append([precision, size, t])
        print('Numba: ', times_nb[-1])
        
        t = timeit.timeit('np.cos(x) + 1j * np.sin(x)', number=100, globals={'x': x, 'np': np})
        times_nptrig.append([precision, size, t])
        print('Trig:  ', times_nptrig[-1])
        
        t = timeit.timeit('np.exp(1j * x)', number=100, globals={'x': x, 'np': np})
        times_npexp.append([precision, size, t])
        print('NPExp: ', times_npexp[-1])
```
Results:
```
Numba:  [<class 'numpy.float32'>, (1000,), 0.0010846489894902334]
Trig:   [<class 'numpy.float32'>, (1000,), 0.0014565419987775385]
NPExp:  [<class 'numpy.float32'>, (1000,), 0.0027067579940194264]
Numba:  [<class 'numpy.float32'>, (100, 100), 0.00136160000693053]
Trig:   [<class 'numpy.float32'>, (100, 100), 0.007243244006531313]
NPExp:  [<class 'numpy.float32'>, (100, 100), 0.024668825004482642]
Numba:  [<class 'numpy.float32'>, (100, 1000), 0.009999030997278169]
Trig:   [<class 'numpy.float32'>, (100, 1000), 0.052969254000345245]
NPExp:  [<class 'numpy.float32'>, (100, 1000), 0.23907336599950213]
Numba:  [<class 'numpy.float32'>, (1000, 1000), 0.10627347399713472]
Trig:   [<class 'numpy.float32'>, (1000, 1000), 0.525297276006313]
NPExp:  [<class 'numpy.float32'>, (1000, 1000), 2.4087961960030952]
Numba:  [<class 'numpy.float32'>, (2, 100, 10000), 0.256061416992452]
Trig:   [<class 'numpy.float32'>, (2, 100, 10000), 1.120946583992918]
NPExp:  [<class 'numpy.float32'>, (2, 100, 10000), 4.803158134003752]
Numba:  [<class 'numpy.float64'>, (1000,), 0.0002444709971314296]
Trig:   [<class 'numpy.float64'>, (1000,), 0.0026107279991265386]
NPExp:  [<class 'numpy.float64'>, (1000,), 0.0032617460092296824]
Numba:  [<class 'numpy.float64'>, (100, 100), 0.0022893300047144294]
Trig:   [<class 'numpy.float64'>, (100, 100), 0.02778685798693914]
NPExp:  [<class 'numpy.float64'>, (100, 100), 0.03169294699910097]
Numba:  [<class 'numpy.float64'>, (100, 1000), 0.008602872010669671]
Trig:   [<class 'numpy.float64'>, (100, 1000), 0.2903215599944815]
NPExp:  [<class 'numpy.float64'>, (100, 1000), 0.32992226698843297]
Numba:  [<class 'numpy.float64'>, (1000, 1000), 0.13833526299276855]
Trig:   [<class 'numpy.float64'>, (1000, 1000), 2.9360293710051337]
NPExp:  [<class 'numpy.float64'>, (1000, 1000), 3.13000461399497]
Numba:  [<class 'numpy.float64'>, (2, 100, 10000), 0.3385424599982798]
Trig:   [<class 'numpy.float64'>, (2, 100, 10000), 5.988608548999764]
NPExp:  [<class 'numpy.float64'>, (2, 100, 10000), 6.3800736170087475]
```

So that's a pretty big win!

**Comment by bmcfee** [comment]
At risk of bike-shedding, what should we call the function which maps angles to unit complex numbers?  I used `phasor` above, but that's probably not quite precise enough.  Some ideas for alternatives:

- `angle_to_c` (or `_to_z`)
- `angle_to_phase`
- `rad_to_c` (or phase, or z) to be more explicit about units
- `unit_phasor`

**Comment by lostanlen** [comment]
i'm fine with any except `angle_to_phase`

`angle_to_z` with a `deg=False` kwarg would be my favorite

**Comment by bmcfee** [comment]
Actually, this makes me wonder if it's worth extending the API a bit to support general phasors (with magnitude scaling).  This would essentially provide an inverse function to `magphase` (though not really, because magphase returns unit phasors as the angle representation).  A default argument of `mag=1` would produce unit phasors, and it might simplify the naming question a bit.

**Comment by lostanlen** [comment]
> extending the API a bit to support general phasors (with magnitude scaling)

@bmcfee would you like to give an example?

**Comment by bmcfee** [comment]
Not much to it: `D = librosa.phasor(angles, mag=S)` would do what you expect.  We could default `mag=1` in which case it produces unit phasors from angles as originally described.

Could look something like:
```python
@numba.vectorize(['complex64(float32)', 'complex128(float64)'], identity=1+0j)
def _phasor_angles(x):
    return np.cos(x) + 1j * np.sin(x)

def phasor(angles, mag=1):
    z = _phasor_angles(x)
    z *= mag
    return z
```

Maybe it's overkill?

**Comment by bmcfee** [comment]
Follow-up:
```python
@numba.vectorize(['complex64(float32)', 'complex128(float64)'], identity=1)
def _phasor_angles(x):
    return np.cos(x) + 1j * np.sin(x)


def phasor1(angles, mag=1):
    z = _phasor_angles(angles)
    z *= mag
    return z

def phasor2(angles, mag=None):
    z = _phasor_angles(angles)
    if mag is not None:
        z *= mag
    return z
```

The idea of the second implementation is to bypass the magnitude scaling unless it's necessary.  Benchmarks show that additional cost is negligible for unitary phasor construction (as we should expect):

```
Angles:  [<class 'numpy.float32'>, (1000,), 0.001663767994614318]
AngMag:   [<class 'numpy.float32'>, (1000,), 0.0013246139860711992]
MagOpt:  [<class 'numpy.float32'>, (1000,), 0.0007913809968158603]
------
Angles:  [<class 'numpy.float32'>, (100, 100), 0.008888109994586557]
AngMag:   [<class 'numpy.float32'>, (100, 100), 0.010941936983726919]
MagOpt:  [<class 'numpy.float32'>, (100, 100), 0.008994483010610566]
------
Angles:  [<class 'numpy.float32'>, (100, 1000), 0.0932460249750875]
AngMag:   [<class 'numpy.float32'>, (100, 1000), 0.10684521897928789]
MagOpt:  [<class 'numpy.float32'>, (100, 1000), 0.09148646600078791]
------
Angles:  [<class 'numpy.float32'>, (1000, 1000), 0.9274552990100347]
AngMag:   [<class 'numpy.float32'>, (1000, 1000), 1.0850995069777127]
MagOpt:  [<class 'numpy.float32'>, (1000, 1000), 0.9439472320082132]
------
Angles:  [<class 'numpy.float32'>, (2, 100, 10000), 1.8817962210159749]
AngMag:   [<class 'numpy.float32'>, (2, 100, 10000), 2.1616768020030577]
MagOpt:  [<class 'numpy.float32'>, (2, 100, 10000), 1.9107237909920514]
------
Angles:  [<class 'numpy.float64'>, (1000,), 0.0016001150070223957]
AngMag:   [<class 'numpy.float64'>, (1000,), 0.0020254630071576685]
MagOpt:  [<class 'numpy.float64'>, (1000,), 0.001694296021014452]
------
Angles:  [<class 'numpy.float64'>, (100, 100), 0.02032391299144365]
AngMag:   [<class 'numpy.float64'>, (100, 100), 0.022394501022063196]
MagOpt:  [<class 'numpy.float64'>, (100, 100), 0.020964575989637524]
------
Angles:  [<class 'numpy.float64'>, (100, 1000), 0.20803841401357204]
AngMag:   [<class 'numpy.float64'>, (100, 1000), 0.2256999370001722]
MagOpt:  [<class 'numpy.float64'>, (100, 1000), 0.21324786799959838]
------
Angles:  [<class 'numpy.float64'>, (1000, 1000), 2.1033617529901676]
AngMag:   [<class 'numpy.float64'>, (1000, 1000), 2.3385029859782662]
MagOpt:  [<class 'numpy.float64'>, (1000, 1000), 2.159378939017188]
------
Angles:  [<class 'numpy.float64'>, (2, 100, 10000), 4.200255053001456]
AngMag:   [<class 'numpy.float64'>, (2, 100, 10000), 4.716209072998026]
MagOpt:  [<class 'numpy.float64'>, (2, 100, 10000), 4.288710249005817]
------
```
`Angles` is `phasor()` from above (no magnitude support, upper-bound for performance); `AngMag` is `phasor1` (always scale), `MagOpt` is `phasor2()` (optional magnitude scaling).

**Comment by bmcfee** [comment]
A quick list of places where `abs2` would be useful:

- `filters.window_bandwidth` https://github.com/librosa/librosa/blob/25538adb3aed3485a06e60b6dad88be3d540f0c2/librosa/filters.py#L1119
- `autocorrelate` https://github.com/librosa/librosa/blob/25538adb3aed3485a06e60b6dad88be3d540f0c2/librosa/core/audio.py#L848
- `rms` https://github.com/librosa/librosa/blob/25538adb3aed3485a06e60b6dad88be3d540f0c2/librosa/feature/spectral.py#L948-L961
- `icqt` https://github.com/librosa/librosa/blob/25538adb3aed3485a06e60b6dad88be3d540f0c2/librosa/core/constantq.py#L717

**Comment by lostanlen** [comment]
> `WINDOW_BANDWIDTHS[key] = n * np.sum(win ** 2) / np.sum(np.abs(win)) ** 2` 

no `abs2` detected ... it's a squared sum, not a sum of squares o__0
but also, i'm not sure how this formula relates with bandwidth? a docstring would be helpful

For the other three, yes 👍🏻 
i did the same exercise of seeking call sites for `abs2` in the librosa call sites and found the same three

**Comment by bmcfee** [comment]
Places where `phasor` would be helpful:

- `filters.wavelet` https://github.com/librosa/librosa/blob/25538adb3aed3485a06e60b6dad88be3d540f0c2/librosa/filters.py#L930-L935
- `phase_vocoder` https://github.com/librosa/librosa/blob/25538adb3aed3485a06e60b6dad88be3d540f0c2/librosa/core/spectrum.py#L1275
- `griffinlim` https://github.com/librosa/librosa/blob/25538adb3aed3485a06e60b6dad88be3d540f0c2/librosa/core/spectrum.py#L2418
- `griffinlim_cqt` https://github.com/librosa/librosa/blob/25538adb3aed3485a06e60b6dad88be3d540f0c2/librosa/core/constantq.py#L1432

**Comment by bmcfee** [comment]
I'll knock this one out early to unblock #1526 for the scipy sprint
